scipy/optimize/_pava/pava_pybind.cpp

```diff
@@ -0,0 +1,115 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <numpy/arrayobject.h>
+
+namespace py = pybind11;
+
+namespace {
+
+intptr_t pava(
+    py::array_t<double, py::array::c_style | py::array::forcecast> xa,
+    py::array_t<double, py::array::c_style | py::array::forcecast> wa,
+    py::array_t<intptr_t, py::array::c_style | py::array::forcecast> ra
+) {
+    // x is the response variable (often written as y). Its ordering is crucial.
+    // Usually, it is sorted according to some other data (feature or covariate), e.g.
+    //   indices = np.argsort(z)
+    //   x = x[indices]
+    // Note that x is modified inplace and, on return, will contain the solution.
+    // w is an array of case weights, modified inplace.
+    // r is an array of indices such that x[r[i]:r[i+1]] contains the i-th block,
+    // modified inplace.
+
+    py::buffer_info x_buffer_info = xa.request();
+    intptr_t n = x_buffer_info.shape[0];
+    double* x = static_cast<double*>(x_buffer_info.ptr);
+    double* w = static_cast<double*>(wa.request().ptr);
+    intptr_t* r = static_cast<intptr_t*>(ra.request().ptr);
```

Comment:
Rather than getting the buffer info and doing the static casts here you can use the proxy objects which also provide some nice nice metadata access. 

```suggestion
    auto x_acc = xa.unchecked_mutable<1>(); 
    size_t n = x_acc.shape(0);
    auto w_acc = wa.unchecked_mutable<1>();
    auto r_acc = ra.unchecked_mutable<1>();
```

Note as the name implies the proxy objects do not implement bounds checking, also the non-type template parameter introduces a compile time generated check that will fail if `xa`,`wa`,`ra` do not have `ndim == 1`.

The existing code below does not use bounds checked access and already implicitly assumes the dimensionality of the input arrays, so I don't see this being an issue.

I am also suggesting the replacement of your use of `intptr_t` with `size_t` as you exclusively use them as offsets and do not cast to pointers anywhere.

Reply:
Also, using the proxy objects the stride information is used to compute offsets so the strict dense packing requirement imposed by `c_style` may be dropped. An argument may be correctly processed if it does not have `stride[0] == 1`.

Reply:
> I am also suggesting the replacement of your use of intptr_t with size_t as you exclusively use them as offsets and do not cast to pointers anywhere.

I'm not 100% sure what the corresponding ndarray dtype for `size_t` is. I tried `np.uintp` and replaced all `intptr_t` by `std::size_t`. But then I get a segmentation fault.

Reply:
It looks like `np.intp` is an alias for `np.int_` which is a platform dependent signed type (64 bit signed int on 64 bit systems) so the corresponding c++ type would be `ssize_t`. That was a minor suggestion though so feel free to switch back.